# **Project 1 - Sales Analysis - ETL**

## Objectives

- Extract data from provided CSV files
- Clean it 
- Apply feature engineering if necessary 
- Remove any unnecessary columns
- Save to a new CSV file


## Inputs

CSV files provided:

stores data-set.csv
sales data-set.csv
features data set.csv

Renamed files to a uniform naming convention:

Sales_Features_DataSet.csv
Sales_DataSet.csv
Sales_Stores_DataSet.csv

Note: original files are stored in Data/OriginalFiles


## Outputs

CSV file created from ETL etc stored in Data/CleanedDataSets:

Sales_DataSet_Cleaned.csv


## Additional Comments

Developed an experimental ETL library which is in this project (modETL_library.py) 
Has lots of cool features so will be interesting to see how it works "in the field"



Used various AI tools to help with the ETL process:
- ChatGPT
- GitHub Copilot. 

See Documents/What_AI_Used_For.md for more details.

## Initalise Working Environment

In [1]:
#import libraries
import os
import numpy as np
import pandas as pd


#below solution provided by chatGPT 
import sys
from pathlib import Path

project_root = Path.cwd().parent

if str(project_root / "assets" / "python_files") not in sys.path:
    sys.path.insert(0, str(project_root / "assets" / "python_files"))
    
import modGlobal
import modETL_Library as modETL
#end solution provided by chatGPT



# Section 1 - Initalisation

## Intitialise All Variables To Be Used In Global Stack 

In [5]:
#DataFrame vars for ETL
dfSales_DataSet = None
dfSales_DataSet_Work = None
dfCleaned = None
dfCheck = None

#missing values check vars
intMissing = 0
intMissingPercent = 0
intLessThanZero = 0
dfCheck = None

#duplicates check vars
intDuplicates = 0

#stores current directory
strCurrentDir = ""

#other vars
objColumns = None
dictDataFrames = dict()

## Set Current Directory To Base Project Directory

In [3]:
#get project directory - default is jupyter notebook sub folder as that is where this file is located!
#so move back one to the project root path

# Source - https://stackoverflow.com/a/17726833
# Posted by chimpsarehungry
# Retrieved 2026-07-05, License - CC BY-SA 3.0

#get current folder
strCurrentDir = os.getcwd()

#is the last part of the path the project directory?
if not strCurrentDir.endswith(modGlobal.CNST_STR_PROJECT_DIR):
   #get current working directory and move back one to the project root path
   strCurrentDir =  os.path.normpath(os.getcwd() + os.sep + os.pardir)
   os.chdir(os.path.dirname(strCurrentDir))
   #change directory
   os.chdir(strCurrentDir)

#confirm current directory is project directory
print(f"Current Directory: \n {os.getcwd()}")

Current Directory: 
 /Users/rogerwilliams/Projects/Python/CourseProjects/Project2_SalesAnalysis


# Rename Raw data files To A Uniform naming Convention

In [4]:
#rename raw files to better names
os.rename(os.getcwd() + "/sales data-set.csv", os.getcwd() + "/Sales_DataSet.csv")
os.rename(os.getcwd() + "/stores data-set.csv", os.getcwd() + "/Stores_DataSet.csv")
os.rename(os.getcwd() + "/features data set.csv", os.getcwd() + "/Features_DataSet.csv")



## Note: purposely leaving the csv files in the root folder for testing purposes
         if project was commercial would delete them after initial processing below

# Section 1 - Extraction

- Read csv file sales data-set.csv as other two do not require ETL (boo!)
- Move into working files directory
- Read csv file into a pandas dataframe
- Get schema info -> column and row numbers
- Get first 5 records
- Get list of column datatypes
- Get detailed schema infomation

## Read csv Files Into Variable For Processing

In [ ]:
#read csv file into DataFrame
dfSales_DataSet = modETL.funcReadFileReturnDataFrame("Sales_DataSet.csv")
#save as working csv file
modETL.funcSaveDataFrameToWorkingFile(dfSales_DataSet)

#now repeat for the other files
dfSales_DataSet = modETL.funcReadFileReturnDataFrame("Stores_DataSet.csv")
#save as working csv file
modETL.funcSaveDataFrameToWorkingFile(dfSales_DataSet)

dfSales_DataSet = modETL.funcReadFileReturnDataFrame("Features_DataSet.csv")
#save as working csv file
modETL.funcSaveDataFrameToWorkingFile(dfSales_DataSet)



#read file from working files folder
#ETL libaray returns a dictionary of all files in the folder with the attribute name
#set to the actual csv filename
dictDataFrames = modETL.funcReadWorkingFilesReturnDictionary()
dfSales_DataSet = dictDataFrames["Sales_DataSet_Working.csv"]

Folder Structure Created Successfully!
Read Sales_DataSet.csv Into DataFrame

Saved: Sales_DataSet To Working Folder
Folder Structure Created Successfully!
Read Stores_DataSet.csv Into DataFrame

Saved: Stores_DataSet To Working Folder
Folder Structure Created Successfully!
Read Features_DataSet.csv Into DataFrame

Saved: Features_DataSet To Working Folder
3 csv Files Read Into DataFrames

DataFrames Created:
Features_DataSet_Working.csv
Stores_DataSet_Working.csv
Sales_DataSet_Working.csv


dict_keys(['Features_DataSet_Working.csv', 'Stores_DataSet_Working.csv', 'Sales_DataSet_Working.csv'])


## Get Schema Info - Sales_DataSet

In [10]:
modETL.funcGetStructure(dfSales_DataSet)
dfSales_DataSet.head()

DataFrame Structure:
<class 'pandas.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Unnamed: 0    421570 non-null  int64  
 1   Store         421570 non-null  int64  
 2   Dept          421570 non-null  int64  
 3   Date          421570 non-null  str    
 4   Weekly_Sales  421570 non-null  float64
 5   IsHoliday     421570 non-null  bool   
dtypes: bool(1), float64(1), int64(3), str(1)
memory usage: 16.5 MB
None


Summary of DataFrame Structure:
Numeric Columns:

          Unnamed: 0          Store           Dept   Weekly_Sales
count  421570.000000  421570.000000  421570.000000  421570.000000
mean   210784.500000      22.200546      44.260317   15981.258123
std    121696.920829      12.785297      30.492054   22711.183519
min         0.000000       1.000000       1.000000   -4988.940000
25%    105392.250000      11.000000      18.000000    2079.650000
50%    

,Unnamed: 0,Store,Dept,Date,Weekly_Sales,IsHoliday
0,0,1,1,05/02/2010,24924.50,False
1,1,1,1,12/02/2010,46039.49,True
2,2,1,1,19/02/2010,41595.55,False
3,3,1,1,26/02/2010,19403.54,False
4,4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet
421,570 rows
6 columns
File size: 16.5 MB

Columns:

Unnamed ,Store, Dept, Date, Weekly_Sales, IsHoliday

Assuming Unnamed is an index column and can be dropped if necessary

We can also see the data types:

Unnamed : int64
Store : int64
Dept : int64
Date : str
Weekly_Sales : float64
IsHoliday : bool

In the Summary of Dataframe Structure for numerical values we see this:

count  421570.000000  421570.000000  421570.000000  421570.000000
mean   210784.500000      22.200546      44.260317   15981.258123
std    121696.920829      12.785297      30.492054   22711.183519
min         0.000000       1.000000       1.000000   -4988.940000
25%    105392.250000      11.000000      18.000000    2079.650000
50%    210784.500000      22.000000      37.000000    7612.030000
75%    316176.750000      33.000000      74.000000   20205.852500
max    421569.000000      45.000000      99.000000  693099.360000

**What Does This Mean?**

- Count is how many values are in the columns
- Mean is the average value of the column
- Std is the standard deviation of the column 
  Standard deviation is how far from the average (mean) the values are. The closer to 0 the
  more consistent the values are i.e. not much deviation from the mean. The deviation range 
  can be referred to as the Sigma. The measurement is based on the Alpha this is a percentage
  of the mean used to determine variance. ypically this is set to 5%
- Min is the minimum value of the column
- 25% is the value at the 25th percentile of the column
- 50% is the value at the 50th percentile of the column 
- 75% is the value at the 75th percentile of the column

**What?**

Percentile is a technical term for a 25% segment of the data. 
50% is the mean, 25% is below that and 75% is 25%above that, this crude picture illustrates this:

 1st Percentile      2nd Percentile        3rd Percentile
0      -      25%  25%      -     50%   75%      -     100%  



Note: Percentile is also referred too as Interquartile Range (IQR)
      Values in 2nd percentile can also be called Central Tendency

- Max is the maximum value of the column




## Get Schema Statistics

In [ ]:
#get column data types
modETL.funcGetStatistics(dfSales_DataSet)

DataFrame Statistics:
          Unnamed: 0          Store           Dept   Weekly_Sales
count  421570.000000  421570.000000  421570.000000  421570.000000
mean   210784.500000      22.200546      44.260317   15981.258123
std    121696.920829      12.785297      30.492054   22711.183519
min         0.000000       1.000000       1.000000   -4988.940000
25%    105392.250000      11.000000      18.000000    2079.650000
50%    210784.500000      22.000000      37.000000    7612.030000
75%    316176.750000      33.000000      74.000000   20205.852500
max    421569.000000      45.000000      99.000000  693099.360000


DataFrame Shape:
Rows: 421570, Columns: 6


Missing Values Per Column:


Duplicate Value Count Per Column:
Store                Total Values: 421570  Duplicate Values: 421525, 99.99% Percent Duplicate Values
Dept                 Total Values: 421570  Duplicate Values: 421489, 99.98% Percent Duplicate Values
Date                 Total Values: 421570  Duplicate Values: 421427, 99.9

## Observations - Sales_DataSet

Date is a string - needs conversion to datetime  
Other column types look ok
If unnamed causes issues later on will drop the column!

Observations:
- No missing values
- High duplicates but this is to be expected as store number/date and department would be repeated


---

# Section 2 - 

- If missing values determine what to fill with (mean, mode or categorical something else)
- Transform data types if necessary

## Get Unique Values - Sales_DataSet

In [10]:
#show unique values count
dfSales_DataSet.nunique()

Store               45
Dept                81
Date               143
Weekly_Sales    359464
IsHoliday            2
dtype: int64

## Observations - Sales_DataSet

Store has 45 unique values
Dept has 81 unique values
Date has 143 unique values
Weekly_Sales has 349,464 unique values
IsHoliday has 2 unique values

## Get Categorical Value Distribution

In [11]:
#get categorical distribution for object columns (non numeric)
for objColumn in ["Date"]:
    print( f"[{objColumn}] Value Distribution:")
    print( dfSales_DataSet[objColumn].value_counts())

[Date] Value Distribution:
Date
23/12/2011    3027
25/11/2011    3021
16/12/2011    3013
09/12/2011    3010
17/02/2012    3007
              ... 
09/07/2010    2903
16/07/2010    2901
20/08/2010    2901
27/08/2010    2898
13/08/2010    2896
Name: count, Length: 143, dtype: int64


## Observations - Sales_DataSet

Nothing obvious is visual as far as extreme deviations

In [12]:
## Create Copy Of DataFrame
dfSales_DataSet_Work = dfSales_DataSet.copy()

## Strip Spaces From Date Column

In [13]:
#remove spaces from date column (string datatype)
for objColumn in["Date"]:
    dfSales_DataSet_Work[objColumn] = dfSales_DataSet_Work[objColumn].str.strip()
    
#show first 5 rows to check    
dfSales_DataSet_Work.head()    

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False
3,1,1,26/02/2010,19403.54,False
4,1,1,05/03/2010,21827.90,False


## Observations - Sales_DataSet

No data corruptions

## Validate Numerical Values

see if Stores column has values less than 1
see if Dept column has values less than 1
see if Weekly_Sales column have values less then 0

In [14]:
#get count of how many values in Stores column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Store"] < 1).sum()
#print result
intLessThanOne

0

---

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Dept Values

In [15]:
#get count of how many values in Dept column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Dept"] < 1).sum()
#print result
intLessThanOne

0

## Observations - Sales_DataSet

Values for Store in acceptable limits

## Validate Weekly_Sales Values

In [16]:
#Get count of how many values in Weekly_Sales column are less than 1
intLessThanOne = (dfSales_DataSet_Work["Weekly_Sales"] < 1).sum()
#print result
intLessThanOne

2006

## Observations - Sales_DataSet

Values for Weekly_Values include negative values
This suggests credits given, will leave data as is


## Data Transformations/Feature Engineering

Date - convert to datetime  

In [17]:
#convert Date to datetime
dfSales_DataSet_Work["Date"] = pd.to_datetime(dfSales_DataSet_Work["Date"], format="%d/%m/%Y")
#show results
dfSales_DataSet_Work.dtypes

Store                    int64
Dept                     int64
Date            datetime64[ns]
Weekly_Sales           float64
IsHoliday                 bool
dtype: object

In [18]:
#double check records
dfSales_DataSet_Work.head()

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,2010-02-05,24924.50,False
1,1,1,2010-02-12,46039.49,True
2,1,1,2010-02-19,41595.55,False
3,1,1,2010-02-26,19403.54,False
4,1,1,2010-03-05,21827.90,False


## Observations - Sales_DataSet

Conversion successful (Strangely Displaying In American Date Foormat??)

## Remove Unnecessary Columns

In [19]:
#nothing to do as no viable columns to remove!


## Save Work DataFrame To CSV File

In [20]:
#create list of columns new and old JUST what is actually needed
lstColumns = [
                #original columns
                "Store","Dept","Date","Weekly_Sales","IsHoliday",
               ]

#create new DataFrame for cleaned data
dfCleaned = dfSales_DataSet_Work[lstColumns].copy()

#save dfSales_DataSet_Work with chatGPT suggested options
dfCleaned.to_csv(strCurrentDir + CNST_STR_CLEANFILEPATH,index=False, date_format="%d/%m/%Y")

#Note: excluded column: Dept as not in the hypothesis analysis which is a nice way
#      avoiding the fact I don't know just how to plot a chart with 98 departments
#      and have it readable!

# Conclusions and Next Steps

ETL went well, DataSet ready for use